File: next-web/lib/adultContent.ts
/** Client-side heuristics matching backend ``is_adult_torrent`` for streamed rows. */

const ADULT_TITLE_PATTERNS: RegExp[] = [
  /\bhentai\b/i,
  /\[18\+\]/i,
  /\(18\+\)/i,
  /\b18\+\b/i,
  /\buncensored\b/i,
  /\bdoujin(?:shi)?\b/i,
  /\bxxx\b/i,
  /\bporn\b/i,
  /\badult\s+only\b/i,
  /\boppai\b/i,
  /\bloli\b/i,
  /\bshotacon\b/i,
  /\bshota\b/i,
];

const NSFW_ENGINE_MARKERS = ["sukebei"];

export function isAdultTorrent(name: string, engineUrl = ""): boolean {
  const title = (name || "").trim();
  if (!title) return false;

  if (ADULT_TITLE_PATTERNS.some((pattern) => pattern.test(title))) {
    return true;
  }

  const engine = (engineUrl || "").toLowerCase();
  return NSFW_ENGINE_MARKERS.some((marker) => engine.includes(marker));
}

export function parseAllowNsfwParam(raw: string | undefined): boolean {
  if (!raw) return false;
  const value = raw.trim().toLowerCase();
  return value === "1" || value === "true" || value === "yes";
}

/** Default is to hide NSFW unless the URL explicitly opts in. */
export function resolveHideNsfw(allowNsfwRaw: string | undefined): boolean {
  return !parseAllowNsfwParam(allowNsfwRaw);
}

File: domain/policies/adult_content.py
"""Heuristics for detecting adult torrent releases in search results."""

from __future__ import annotations

import re

# Case-insensitive markers commonly found in adult torrent titles.
_ADULT_TITLE_PATTERNS: tuple[re.Pattern[str], ...] = tuple(
    re.compile(pattern, re.IGNORECASE)
    for pattern in (
        r"\bhentai\b",
        r"\[18\+\]",
        r"\(18\+\)",
        r"\b18\+\b",
        r"\buncensored\b",
        r"\bdoujin(?:shi)?\b",
        r"\bxxx\b",
        r"\bporn\b",
        r"\badult\s+only\b",
        r"\boppai\b",
        r"\bloli\b",
        r"\bshotacon\b",
        r"\bshota\b",
    )
)

_NSFW_ENGINE_MARKERS: tuple[str, ...] = ("sukebei",)


def is_adult_torrent(name: str, engine_url: str = "") -> bool:
    """Return True when a torrent row looks like NSFW/hentai content."""
    title = (name or "").strip()
    if not title:
        return False

    for pattern in _ADULT_TITLE_PATTERNS:
        if pattern.search(title):
            return True

    engine = (engine_url or "").lower()
    return any(marker in engine for marker in _NSFW_ENGINE_MARKERS)
